File: heap/heapsort.cpp
/* C++ program to perform heapsort */
#include <bits/stdc++.h>
using namespace std;

/* Display the given array */
void display(int *arr, int size)
{
	for (int i = 1; i <= size; i++)
		cout << arr[i] << " ";
	cout << "\n";
}

/* Return left child of given node */
int leftChild(int p)
{
	return (p * 2);
}

/* Return right child of given node */
int rightChild(int p)
{
	return (p * 2 + 1);
}

/* Perform max heapification at given node */
void maxHeapify(int * arr, int i, int size)
{
	int largest = i;
	int left = leftChild(i);
	int right = rightChild(i);

	if (left <= size && arr[left] > arr[largest])
		largest = left;
	if (right <= size && arr[right] > arr[largest])
		largest = right;
	if (largest != i)
	{
		swap(arr[i], arr[largest]);
		maxHeapify(arr, largest, size);
	}
}

/* Build the given array into heap */
void buildHeap(int *arr, int size)
{
	for (int i = floor(size / 2); i >= 1; i--)
		maxHeapify(arr, i, size);
}

/* Sort the given array */
void heapSort(int *arr, int size)
{
	int length = size;
	buildHeap(arr, size);
	for (int i = length; i >= 2; i--)
	{
		swap(arr[1], arr[i]);
		size -= 1;
		maxHeapify(arr, 1, size);
	}
}

/* Begin main */
int main()
{
	int size, arr[100000];

	cout << "Enter heap size.\n";
	cin >> size;

	cout << "Enter heap elements.\n";
	for (int i = 1; i <= size; i++)
		cin >> arr[i];

	heapSort(arr, size);
	display(arr, size);

	return 0;
}
/* End main */

File: heap/heapsort.hs
-- Haskell program to perform heapsort
import Data.Graph.Inductive.Internal.Heap(
  Heap(..),insert,findMin,deleteMin)

-- Build the given list into heap
build :: Ord a => [(a,b)] -> Heap a b  -- build takes a list of (a,b) pairs, returns a heap
build = foldr insert Empty  -- 

toList :: Ord a => Heap a b -> [(a,b)]  -- toList takes a heap and returns a list of ordered tuples
toList Empty = []
toList h = x:toList r
           where (x,r) = (findMin h,deleteMin h)

-- Sort the given list
heapsort :: Ord a => [a] -> [a]  -- heapsort takes a list and returns a list
heapsort = (map fst) . toList . build . map (\x->(x,x))  {- takes the list, creates a pair (x,x) for each element, calls build, 
converts returned heap to list, and applies fst to each element of the heap. fst returns a list of the first elements of each 
tuple in the given list -}

-- Begin main
main=do
	input<-getLine  -- get input
	let n=(read input)::[Int]  -- convert input to list of integers
	heapsort n  -- call heapsort function on input list
	return()
-- End main
